chia/server/ws_connection.py

```diff
@@ -222,10 +222,16 @@ async def perform_handshake(
             if inbound_handshake.network_id != network_id:
                 raise ProtocolError(Err.INCOMPATIBLE_NETWORK_ID)
 
+            if inbound_handshake.protocol_version != protocol_version:
```

Comment:
Is this comparing the right protocol version, even if you're a Wallet and the peer is a FullNode for example? It should be comparing expected wallet version vs actual wallet version, expected full node version vs actual full node version, rather than e.g. expected wallet version vs actual full node version. If that makes sense

Reply:
`protocol_version` is passed in as an argument to this function, `perform_handshake()`. We call this from `chia/server/server.py`, and in both places we pass in the version for our protocol.

See:
https://github.com/Chia-Network/chia-blockchain/pull/17791/files#diff-926449f427065cf100c0af5785ac61529c768ab4cccbef25ccb86e82074b753fR490

https://github.com/Chia-Network/chia-blockchain/pull/17791/files#diff-926449f427065cf100c0af5785ac61529c768ab4cccbef25ccb86e82074b753fR337

Reply:
I'm not following, assume the following scenario:
- machine a with a full node and the following protocol versions
  - full node <-> full node protocol (full node protocol) is 0.0.42
  - wallet <-> full node protocol (wallet protocol) is 0.0.36
- machine b with a wallet and the following protocol versions:
  - full node <-> full node protocol (full node protocol) is 0.0.42
  - wallet <-> full node protocol (wallet protocol) is 0.0.36
- the wallet from machine b connects to the full node on machine a

right now the protocol version of the sent handshake from the wallet would be 0.0.36 (wallet protocol) while the full node responds with its local type protocol version (full node protocol version), 0.0.42, instead of the wallet protocol version.

Is this intended?

Reply:
If you have an incoming connection i'd expect the peer type's protocol version to be used, while if we initiate the connection the local type is used

Reply:
I will review this logic in more detail. thanks for pointing this out.

Reply:
I made a change where, when accepting an incoming connection, the protocol version in the handshake is the version for the same protocol as the incoming node type. I'm starting to realize that the "node type" is slightly different from "protocol", even though they seem the same. there's a type for full node, and a full node protocol, a type for wallet, and a wallet protocol.
But the way this works now is that we always say what kind of *service* we are, not which protocol we're talking to you with. It's a little bit strange.

Reply:
Yes, i agree